### PR TITLE
Minesweeper: Faster generation of game fields

### PR DESCRIPTION
**Pull request by kuzux**

### CODE REVIEW

File: Userland/Games/Minesweeper/Field.cpp [hunk 1]
```diff
@@ -269,6 +264,57 @@ void Field::reset()
...
+        m_squares[i]->has_mine = false;
+        m_squares[i]->has_flag = false;
+        m_squares[i]->is_considering = false;
+        m_squares[i]->is_swept = false;
+        m_squares[i]->number = 0;
+        if (!free_squares.contains(i))
+            possible_mine_positions.unchecked_append(i);
+    }
+
+    // Fisher-Yates shuffle
```
Comment (AtkinsSJ) [inline Userland/Games/Minesweeper/Field.cpp:301]: This same shuffle is in `Userland/Utilities/shuf.cpp`. It would be nice to move it into `AK/Random.h` instead of duplicating it.

File: Userland/Games/Minesweeper/Field.cpp [hunk 2]
```diff
@@ -269,6 +264,57 @@ void Field::reset()
...
+    size_t tmp;
+    for (size_t i = possible_mine_positions.size() - 1; i >= 1; --i) {
+        size_t j = get_random_uniform(i + 1);
+        // Swap i and j
+        if (i == j)
+            continue;
+        tmp = possible_mine_positions[j];
+        possible_mine_positions[j] = possible_mine_positions[i];
+        possible_mine_positions[i] = tmp;
```
Comment (AtkinsSJ) [inline Userland/Games/Minesweeper/Field.cpp:310]: ...while moving it, we also have an `AK::swap(a, b)` function in `AK/StdLibExtras.h`

File: Userland/Games/Minesweeper/Field.cpp [hunk 3]
```diff
@@ -278,14 +324,14 @@ void Field::reset()
             });
             square.number = number;
             if (square.has_mine)
-                continue;
-            if (square.number)
+                square.label->set_icon(m_mine_bitmap);
+            else if (square.number) {
                 square.label->set_icon(m_number_bitmap[square.number - 1]);
+            }
```
Comment (AtkinsSJ) [inline Userland/Games/Minesweeper/Field.cpp:330]: These changes seem unrelated to the rest of this commit, unless I'm missing something.

File: Userland/Games/Minesweeper/Field.cpp [hunk 4]
```diff
@@ -278,14 +324,14 @@ void Field::reset()
...
         }
     }
 
     m_unswept_empties = rows() * columns() - m_mine_count;
-    set_updates_enabled(true);
```
Comment (AtkinsSJ) [inline Userland/Games/Minesweeper/Field.cpp:288]: ...same for this line.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
I have some suggestions. Overall this works really well though, it's basically instant even on maximum settings! Though a maximum-mines game is not especially fun to play. :wink:

File: Userland/Games/Minesweeper/Field.cpp (hunk 4 above)
Comment (kuzux) [reply Userland/Games/Minesweeper/Field.cpp:288]: This one's just due to how diffing worked, there's another line in the diff below adding it back :D

File: Userland/Games/Minesweeper/Field.cpp (hunk 3 above)
Comment (kuzux) [reply Userland/Games/Minesweeper/Field.cpp:330]: This one was fixing a bug introduced when the generation was changed, would it be better if the fix was in a separate commit?

File: Userland/Games/Minesweeper/Field.cpp (hunk 4 above)
Comment (AtkinsSJ) [reply Userland/Games/Minesweeper/Field.cpp:288]: Ahhh, you're right, I missed that. git is weird sometimes. :sweat_smile:
